### ISSUE TITLE
MemoryLeak - RCTCxxBridge dispatchBlock:queue:

### ISSUE THREAD

**Issue by ajwhite** [facebook/react-native#18043]
## TL;DR

`setTimeout` leads to endless growth in `RCTCxxBridge`:

<img width="504" alt="screen shot 2018-02-21 at 11 51 37 am" src="https://user-images.githubusercontent.com/656630/36493435-9d969d80-16fd-11e8-8b5c-aff61d943053.png">

### Is this a bug report?

Yes

### Have you read the [Contributing Guidelines](https://facebook.github.io/react-native/docs/contributing.html)?

Yes

### Environment

```
Environment:
  OS: macOS High Sierra 10.13.2
  Node: 6.10.2
  Yarn: 0.24.5
  npm: 3.10.10
  Watchman: 4.7.0
  Xcode: Xcode 9.2 Build version 9C40b
  Android Studio: 3.0 AI-171.4443003

Packages: (wanted => installed)
  react: 16.2.0 => 16.2.0
  react-native: 0.53.3 => 0.53.3
```

### Steps to Reproduce

1. Create a timer
2. Run the app in release mode
3. Observe that `Malloc 64 Bytes` instances are retained

### Expected Behavior

Memory should not grow over time.

### Actual Behavior

Every time `setTimeout` either runs, or becomes queue'd (hard to tell), it leads to a leak.

One interesting behavior is that it climbs, levels out, and then climbs again later.

<img width="504" alt="screen shot 2018-02-21 at 11 51 37 am" src="https://user-images.githubusercontent.com/656630/36493525-d1965dfa-16fd-11e8-8738-aa6d51a3e3d8.png">

<img width="883" alt="screen shot 2018-02-21 at 11 51 28 am" src="https://user-images.githubusercontent.com/656630/36493524-d1875576-16fd-11e8-926b-57b70abc4272.png">

### Reproducible Demo

https://github.com/ajwhite/react-native-memory-leak-nsdate

Create an endless timer and observe the memory growth. (https://github.com/ajwhite/react-native-memory-leak-nsdate/commit/c94205def099e5da01bedfe7f1bd4c0b14200ca7)

```js
// somewhere in your JS:

function timerLeak() {
  setTimeout(timerLeak, 100)
}

timerLeak()
```

Then open up Instruments and observe memory allocations under `Malloc 64 Byes`

Related to #18042
